### PR TITLE
Makes corrections to QueryPlanVisitors for dealing with DeltaDataSource

### PR DESCRIPTION
**Pull request by collado-mike**
### Problem
Address both https://github.com/OpenLineage/OpenLineage/issues/458 and https://github.com/OpenLineage/OpenLineage/issues/459 . This corrects the Hadoop Configuration construction for underlying LogicalRelations and also adds specific handling for Delta relations to the `SaveIntoDataSourceCommandVisitor`.  

As part of the latter fix, I needed to tweak how the tests manage the permits for the test execution. The Delta source ends up triggering multiple events, which need to run simultaneously, so multiple open permits became necessary. I picked `5` as a random number of permits to support. It's arbitrary and can be changed if needed.

Closes: #458 and #459

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [X] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [X] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [X] Your changes are accompanied by tests (_if relevant_)
- [X] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [X] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)

### CODE REVIEW

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/OpenLineageSparkListener.java [hunk 1]
```diff
@@ -274,6 +275,12 @@ private static void clear() {
     outputs.clear();
   }
 
+  @Override
+  public void onApplicationEnd(SparkListenerApplicationEnd applicationEnd) {
+    close();
+    super.onApplicationEnd(applicationEnd);
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/OpenLineageSparkListener.java:281]: Was this a bug that kept connections / resources open?

File: integration/spark/src/test/common/java/io/openlineage/spark/agent/lifecycle/StaticExecutionContextFactory.java [hunk 2]
```diff
@@ -16,39 +16,57 @@
...
 /** Returns deterministic fields for contexts */
 public class StaticExecutionContextFactory extends ContextFactory {
-  public static final Semaphore semaphore = new Semaphore(1);
+
+  // Create a semaphore with multiple permits so that jobs that launch multiple Spark SQL jobs,
+  // (i.e., Delta jobs that need to work with Delta logs as well as the target dataset) and fire
+  // multiple SparkListenerSQLExecutionStart events don't block waiting for the first SQL job to
+  // finish. The #waitForExecutionEnd method will wait for <i>all</i> acquired permits to be
+  // released before continuing.
+  public static final int NUM_PERMITS = 5;
```
Comment (wslulciuc) [inline integration/spark/src/test/common/java/io/openlineage/spark/agent/lifecycle/StaticExecutionContextFactory.java:39]: _Minor_: Maybe chose a multiple of `2`? i.e. `4` or `8`?

File: integration/spark/src/test/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/DeltaDataSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+@ExtendWith(SparkAgentTestExtension.class)
+public class DeltaDataSourceTest {
+
+  @Test
+  public void testInsertIntoDeltaSource(@TempDir Path tempDir, SparkSession spark)
```
Comment (wslulciuc) [inline integration/spark/src/test/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/DeltaDataSourceTest.java:36]: Solid test! 💯

**Review by wslulciuc — APPROVED** [review]
Left some minor comments, otherwise 👍

**Comment by wjohnson** [comment]
Just adding my two cents here. The fixes worked great on our testing inside of Azure Databricks.  Thank you @collado-mike for your amazing work as usual!

**Pull request merged**